File: java/2-finish/akka-streams/src/main/java/pipelines/workshop/egress/CardPaymentLogger.java
package pipelines.workshop.egress;

import akka.NotUsed;
import akka.stream.javadsl.RunnableGraph;
import akka.stream.javadsl.Sink;
import pipelines.akkastream.AkkaStreamlet;
import pipelines.akkastream.javadsl.RunnableGraphStreamletLogic;
import pipelines.streamlets.StreamletShape;
import pipelines.streamlets.avro.AvroInlet;
import pipelines.workshop.schema.java.CardPayment;

public class CardPaymentLogger extends AkkaStreamlet {

    AvroInlet<CardPayment> inlet = AvroInlet.<CardPayment>create("in", CardPayment.class);

    @Override
    public StreamletShape shape() {
        return StreamletShape.createWithInlets(inlet);
    }

    public RunnableGraphStreamletLogic createLogic() {
        return new RunnableGraphStreamletLogic(getStreamletContext()) {
            @Override
            public RunnableGraph<NotUsed> createRunnableGraph() {
                return getAtMostOnceSource(inlet).to(Sink.foreach(payment -> System.out.println("Hello, Card Payment! => " + payment)));
            }
        };
    }
}

File: scala/2-finish/akka-streams/src/main/scala/pipelines/workshop/egress/CardPaymentLogger.scala
package pipelines.workshop.egress

import akka.stream.scaladsl.RunnableGraph
import pipelines.akkastream.scaladsl.{ FlowWithPipelinesContext, RunnableGraphStreamletLogic }
import pipelines.akkastream.{ AkkaStreamlet, StreamletLogic }
import pipelines.streamlets.StreamletShape
import pipelines.streamlets.avro.AvroInlet
import pipelines.workshop.schema.CardPayment

object CardPaymentLogger extends AkkaStreamlet {

  val inlet = AvroInlet[CardPayment]("in")
  val shape = StreamletShape.withInlets(inlet)

  override protected def createLogic(): StreamletLogic = new RunnableGraphStreamletLogic() {

    val flow = FlowWithPipelinesContext[CardPayment].map { cp ⇒
      system.log.info("Hello, Card Payment! " + cp.toString)
      cp
    }

    override def runnableGraph(): RunnableGraph[_] =
      atLeastOnceSource(inlet).via(flow).to(atLeastOnceSink)
  }
}
